coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -436,42 +432,64 @@ private static boolean isUnionSchemaCompatible(@Nonnull Schema leftSchema, @Nonn
-            && isUnionSchemaCompatible(leftSchema.getValueType(), rightSchema.getValueType(), strictMode);
+        if (leftSchema.getType() != rightSchema.getType()) {
+          return null;
+        }
+        Schema valueType = getUnionFieldSchema(leftSchema.getValueType(), rightSchema.getValueType(), strictMode);
+        return valueType != null ? Schema.createMap(valueType) : null;
 
       case ARRAY:
-        return leftSchema.getType() == rightSchema.getType()
-            && isUnionSchemaCompatible(leftSchema.getElementType(), rightSchema.getElementType(), strictMode);
+        if (leftSchema.getType() != rightSchema.getType()) {
+          return null;
+        }
+        Schema elementType = getUnionFieldSchema(leftSchema.getElementType(), rightSchema.getElementType(), strictMode);
+        return elementType != null ? Schema.createArray(elementType) : null;
 
       case UNION:
-        boolean isSameUnionType = (leftSchema.getType() == rightSchema.getType());
-        boolean isBothNullableType =
-            AvroSerdeUtils.isNullableType(leftSchema) && AvroSerdeUtils.isNullableType(rightSchema);
-
-        Schema leftOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(leftSchema);
-        Schema rightOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(rightSchema);
-        boolean isOtherTypeUnionCompatible = isUnionSchemaCompatible(leftOtherType, rightOtherType, strictMode);
-
-        return isSameUnionType && isBothNullableType && isOtherTypeUnionCompatible;
-
+        if (leftSchema.getType() == rightSchema.getType()) {
+          boolean isBothNullableType =
+              AvroSerdeUtils.isNullableType(leftSchema) && AvroSerdeUtils.isNullableType(rightSchema);
+
+          Schema leftOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(leftSchema);
+          Schema rightOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(rightSchema);
+          final Schema unionFieldSchema = getUnionFieldSchema(leftOtherType, rightOtherType, strictMode);
+          return isBothNullableType && unionFieldSchema != null
+              ? Schema.createUnion(Arrays.asList(Schema.create(Schema.Type.NULL), unionFieldSchema)) : null;
+        } else {
+          if (!strictMode && AvroSerdeUtils.isNullableType(leftSchema)) {
+            Schema leftOtherType = AvroSerdeUtils.getOtherTypeFromNullableType(leftSchema);
+            return rightSchema.getType() == Schema.Type.NULL
+                || getUnionFieldSchema(leftOtherType, rightSchema, strictMode) != null ? leftSchema : null;
+          }
+          return null;
+        }
+      case NULL:
+        if (rightSchema.getType() == Schema.Type.UNION) { // this case has been handled in UNION above
+          return null;
+        } else if (rightSchema.getType() == Schema.Type.NULL) {
+          return leftSchema;
+        } else {
+          return strictMode ? null : Schema.createUnion(Arrays.asList(leftSchema, rightSchema));
+        }
```

Comment:
Does it make sense to simplify this code by checking whether either `rightSchema` or `leftSchema` is `NULL` at the beginning of the method, then modifying `rightSchema` and `leftSchema` accordingly (e.g., make both `[null, other-type]`, then we we can proceed with the original function (the version before this patch)? cc: @funcheetah.

Reply:
If we choose to do that, I think the code will be more complex than current patch, because this patch just modifies the return value of 2 functions (`isUnionRecordSchemaCompatible` and `isUnionSchemaCompatible`) from `boolean` to union result. 
But we need to add another complex function to modify two schemas apart from 2 boolean functions if we choose to keep `isUnionRecordSchemaCompatible` and `isUnionSchemaCompatible`.

Reply:
merge union schema while traverse the schema tree seems cleaner. Otherwise, we need to traverse the schema tree again since null type can be nested.